include/samurai/algorithm/update.hpp

```diff
@@ -817,6 +818,170 @@ namespace samurai
 
     template <class Tag, class Field, class... Fields>
     bool update_field_mr(const Tag& tag, Field& field, Fields&... other_fields)
+    {
+        using mesh_t                     = typename Field::mesh_t;
+        static constexpr std::size_t dim = mesh_t::dim;
+        using mesh_id_t                  = typename Field::mesh_t::mesh_id_t;
+        using size_type                  = typename Field::size_type;
+        using interval_t                 = typename mesh_t::interval_t;
+        using value_t                    = typename interval_t::value_t;
+        using unsigned_value_t           = typename std::make_unsigned_t<value_t>;
+        using lca_type                   = typename Field::mesh_t::lca_type;
+        using lca_const_iterator         = typename lca_type::const_iterator;
+        using ca_type                    = typename Field::mesh_t::ca_type;
+        using coord_type                 = typename lca_type::coord_type;
+
+        const auto& mesh = tag.mesh();
+
+#ifdef SAMURAI_WITH_MPI
+        constexpr bool useMPI = true;
+#else
+        constexpr bool useMPI = false;
+#endif // SAMURAI_WITH_MPI
+
+        ca_type ca_add_m;
+        ca_type ca_add_p;
+        ca_type ca_remove_m;
+        ca_type ca_remove_p;
+        ca_type new_ca;
+
+        std::vector<value_t> add_p_x;
+        std::vector<coord_type> add_p_yz;
+
+        std::vector<size_t> add_p_idx;
+
+        for (std::size_t level = mesh[mesh_id_t::cells].min_level(); level <= mesh[mesh_id_t::cells].max_level(); ++level)
+        {
+            const lca_const_iterator begin = mesh[mesh_id_t::cells][level].cbegin();
```

Comment:
use `auto` instead and remove the type definition at the beginning.

In the following, you also have a lot of type defined. Can you try to use as much as possible the `auto` keyword.